### ISSUE TITLE
Let the user interact with the underlying GHCi by allowing them to type

### ISSUE THREAD

**Issue by Wizek** [ndmitchell/ghcid#129]
I find myself occasionally having to launch two GHCi sessions side by side: one through GHCid, and one without. The latter one I find necessary when I want to manually poke around with the functions of the codebase at hand.

Which feels a bit wasteful when each `ghc -i` process can take up `1-3` GB of RAM.

Wouldn't it be nice if we could interact with the underlying ghci through ghcid just as if it was a regular interactive session? Meaning inputs could be line-buffered-ly passed to the child process, and outputs are already passed back up, AFAIU.

And when a file is changed that ghcid is watching, then the session would immediately reload (running `--test`s again as well if they are specified) just like it does today.

@ndmitchell What do you think of this feature?

**Comment by ndmitchell** [comment]
There are a few reasons I'm not convinced:

* Currently where possible people should be using `-fno-code` when running through `ghcid`. You can't do that and have workable interactions.
* What if you want to change the in-scope module when using that `ghci`? It's not going to give a fully fledged experience.
* You aren't going to get things like auto-complete.
* It seems technically quite difficult, or at least complex - there are many moving parts and things like line buffering etc are a pain.

The benefit seems to have saving a few Gb, but that isn't a huge amount nowadays - I personally just run ghci and ghcid separately. That said, it's not a definitive no, and I have been tempted down this route several times.

**Comment by Wizek** [comment]
Let me address your points out of order:

- 3\. Yes, this is indeed a difficulty I ran into as well when, before hearing about `ghcid`s capabilities, I attempted to write a simple ghci-auto-reloader. Though maybe it could be possible in light of my next point...
- 4\. What, then, if we would have a more raw connection between the parent's and the child processes stdin and stdout? Perhaps we could pipe through each character as they are pressed, even tabs? Quite like how `ssh`, `tmux` and other terminal emulators(?) are able to handle tab completions too. That way, with a lower level solution, we could have a 'higher level', more seamless experience. It could really feel to the end user like using `ghci` + reload while `ghcid` is running.  
I presume some research would be required to find out what API would allow for this &mdash; it may not be as simple as `hGetChar` and `hPutChar`.
- 1\. I am a bit surprised to hear this. Is this an official recommendation? Do many people do this? I almost never do, as I like to run `Spec.main` and/or `Main.main` with `--test` upon file changes. But even still, only allowing `--interactive`, conditionally in disjunction with `--no-code` would be a reasonable and understandable concession to this new feature I think.
- 2\. This is the point I saved for last, because I understand this the least. How come? What would happen if I were to type `import M`, `:m + M`, or `:load M.hs`? Can I somehow try this out, do you think? Do you know some way with which I could inject some lines into the child ghci?  
But even if it were to turn out that module loading would run into some challenges, I think a limited feature that has only access to the current scope (which I think in my apps is usually most if not all of the modules in the package anyway) would be a desirable first step.

What do you think about my responses?

**Comment by ndmitchell** [comment]
Yep, it's certainly possible - I think `hGetChar` and `hPutChar` are all you need - but that doesn't make it simple.

* 1. It's how I use it, and yes, you could disable it when asking for an interactive mode.
* 2. GHCid is pretty sensitive - it has to configure lots of settings like no break on exceptions, and set the prompt etc. Doing things like loading different files will change what is in scope and how ghci responds, and ghcid needs the responses from ghci to be predictable so it can parse them.

I'm not going to close the ticket, but I suspect it's a lot of work with quite a few downsides (as well as some upsides).
